### PR TITLE
#172: Java 8 date and time support

### PR DESCRIPTION
**Pull request by vadeg**
## Issue link
        #172 Java 8 date and time support.

## Purpose
Support Java 8 date and time automated mapping [JSR-310](https://jcp.org/en/jsr/detail?id=310). Converters added for:
* `LocalDate`
* `LocalDateTime`
* `OffsetDateTime`
* `ZonedDateTime`
* `Instant`

Added conversion between types above and
* to and from `String`
* to and from `Date`

## Open Questions and Pre-Merge TODOs
- [x] Issue created
- [ ] Unit tests pass
- [ ] Documentation updated
- [ ] Travis build passed

### CODE REVIEW

**Comment by garethahealy** [comment]
@vadeg ; can you rebase/squash. have pushed license update.

**Comment by garethahealy** [comment]
Links to:
- https://github.com/DozerMapper/dozer/issues/172

File: core/src/main/java/org/dozer/converters/DateFormatContainer.java [hunk 1]
```diff
@@ -17,34 +17,57 @@
...
-  private DateFormat determineDateFormat() {
-    return dfStr == null ? null : new SimpleDateFormat(dfStr, Locale.getDefault());
-  }
+    /**
+     * TODO replace method call with reflection in tests.
+     * @param dateFormat dateFormat
+     * @deprecated This method will break internal state of the instance by setting formatter which is not using
+     * format provided using constructor. It will be removed in future releases.
+     */
+    @Deprecated
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/DateFormatContainer.java]: whys it been deprecated?

File: core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        } else {
+            return super.convert(destClass, srcObject);
+        }
+
+        if (LocalTime.class.isAssignableFrom(destClass)) {
+            return result.toLocalTime();
+        } else if (LocalDate.class.isAssignableFrom(destClass)) {
+            return result.toLocalDate();
+        }
+        return result;
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java]: the return of the result is hard to follow code wise. either return once, or return once result is retrieved.

File: core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        } else if (Instant.class.isAssignableFrom(srcObjectClass)) {
+            result = LocalDateTime.ofInstant(((Instant) srcObject), ZoneId.systemDefault());
+        } else if (Long.class.isAssignableFrom(srcObjectClass)) {
+            Instant instant = Instant.ofEpochMilli((Long) srcObject);
+            result = LocalDateTime.ofInstant(instant, ZoneId.systemDefault());
+        } else {
+            return super.convert(destClass, srcObject);
+        }
+
+        if (LocalTime.class.isAssignableFrom(destClass)) {
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java]: how can this block of code be executed? wouldn't the else always kick-in.

File: core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            Instant instant = Instant.ofEpochMilli((Long) srcObject);
+            result= OffsetDateTime.ofInstant(instant, ZoneId.systemDefault());
+        } else {
+            return super.convert(destClass, srcObject);
+        }
+
+        if (OffsetTime.class.isAssignableFrom(destClass)) {
+            return result.toOffsetTime();
+        }
+        return result;
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java]: same as other, return once or instantly.

File: core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        } else if (Instant.class.isAssignableFrom(srcObjectClass)) {
+            result = OffsetDateTime.ofInstant(((Instant) srcObject), ZoneId.systemDefault());
+        } else if (Long.class.isAssignableFrom(srcObjectClass)) {
+            Instant instant = Instant.ofEpochMilli((Long) srcObject);
+            result= OffsetDateTime.ofInstant(instant, ZoneId.systemDefault());
+        } else {
+            return super.convert(destClass, srcObject);
+        }
+
+        if (OffsetTime.class.isAssignableFrom(destClass)) {
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java]: again. how does the code get here? the else would kick in

File: core/src/main/java/org/dozer/converters/StringConverter.java [hunk 6]
```diff
@@ -16,42 +16,54 @@
...
+    /**
+     * Whether date format is provided or not
+     * @return true - date format provided. Otherwise false.
+     */
+    private boolean hasDateFormat() {
+        return Optional
+                .ofNullable(dateFormatContainer)
+                .map(DateFormatContainer::getDateFormat)
+                .isPresent();
+    }
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/StringConverter.java]: although its nice to use java8 optionals. sometimes its easier to read old fashioned if null checks.

File: core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Internal converter to {@link LocalDateTime} type. Only intended for internal use.
+ * <pre>
+ * Supported conversions:
+ *  Any child of TemporalAccessor &#45;&gt; LocalDateTime ()
+ *  String &#45;&gt; LocalDateTime
+ *  Long &#45;&gt; LocalDateTime
+ *  Date &#45;&gt; LocalDateTime
+ * </pre>
+ * @author vadeg
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java]: remove author tags. git shows history

File: core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import java.time.OffsetTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+
+/**
+ * Internal class for converting supported types to {@link java.time.OffsetDateTime}.
+ * Only intended for internal use.
+ *
+ * @author vadeg
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java]: remove author tags. git shows history

File: core/src/main/java/org/dozer/converters/ZonedDateTimeConverter.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+
+/**
+ * Internal class for converting supported types to {@link java.time.ZonedDateTime}.
+ * Only intended for internal use.
+ *
+ * @author vadeg
```
Comment (garethahealy) [inline core/src/main/java/org/dozer/converters/ZonedDateTimeConverter.java]: remove author tags. git shows history

File: bom-dependencies/pom.xml [hunk 10]
```diff
@@ -279,6 +280,11 @@
                 <artifactId>objenesis</artifactId>
                 <version>${objenesis.version}</version>
             </dependency>
+            <dependency>
+                <groupId>pl.pragmatists</groupId>
+                <artifactId>JUnitParams</artifactId>
+                <version>${junit-params.version}</version>
+            </dependency>
```
Comment (garethahealy) [inline bom-dependencies/pom.xml]: since no other unit tests use this, i'd prefer to not introduce. it creates a strange style to the codebase, i.e.:

- 1 person uses this
- another uses standard junit
- another wants to introduce TestNG

etc etc etc

It doesnt look like it adds a massive amount of benefit, other than not having to call the method at test start up to get some test data.

File: core/src/main/java/org/dozer/converters/DateFormatContainer.java (hunk 1 above)
Comment (vadeg) [reply core/src/main/java/org/dozer/converters/DateFormatContainer.java]: There is a comment above.

File: core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java (hunk 3 above)
Comment (vadeg) [reply core/src/main/java/org/dozer/converters/LocalDateTimeConverter.java]: `LocalDateTime -> LocalTime` conversion will execute this code.

File: core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java (hunk 5 above)
Comment (vadeg) [reply core/src/main/java/org/dozer/converters/OffsetDateTimeConverter.java]: `OffsetDateTime -> OffsetTime` conversion will execute this code.

File: bom-dependencies/pom.xml (hunk 10 above)
Comment (vadeg) [reply bom-dependencies/pom.xml]: Style remains the same. It is still JUnit with small extension. This is not a TestNG of even JUnit5.

It simplifies code a lot. Instead of writing tests with names like `testTimestampConversion`, `testSqlDateConversion`, `testUtilDateConversion` and etc. with the same code except one line it can be rewritten in one test with large amount of test cases. All of these 
* `DateConverterTest`
* `IntegerConverterTest`
* `PrimitiveOrWrapperConverterTest`
* `StringConverterTest`

can be replaced with one test and set of different test cases what will reduce amount of boilerplate code in tests and will make addition of new test cases much more easier than now.

Comment (garethahealy) [reply bom-dependencies/pom.xml]: @vadeg ; sorry, still not convinced. i understand that it reduces code, but it introduces a test dependancy that i've never seen before (i work with a lot of different companies as a job) and only ever used in this 1 class.

**Comment by garethahealy** [comment]
Closed in favour of #685
